physical/etcd.go

```diff
@@ -167,3 +179,188 @@ func (b *EtcdBackend) nodePath(key string) string {
 func (b *EtcdBackend) nodePathDir(key string) string {
 	return filepath.Join(b.path, key) + "/"
 }
+
+// nodePathLock returns an etcd directory path used specifically for semaphore
+// indicies based on the given key.
+func (b *EtcdBackend) nodePathLock(key string) string {
+	return filepath.Join(b.path, filepath.Dir(key), EtcdNodeLockPrefix+filepath.Base(key)+"/")
+}
+
+// Lock is used for mutual exclusion based on the given key.
+func (c *EtcdBackend) LockWith(key, value string) (Lock, error) {
+	return &EtcdLock{
+		backend: c,
+		key:     key,
+		value:   value,
+	}, nil
+}
+
+// EtcdLock emplements a lock using and etcd backend.
+type EtcdLock struct {
+	backend                  *EtcdBackend
+	key, value, semaphoreKey string
+	lock                     sync.Mutex
+	isHeld                   bool
+	done                     chan struct{}
+}
+
+// addSemaphoreKey aquires a new ordered semaphore key.
+func (c *EtcdLock) addSemaphoreKey() (string, uint64, error) {
+
+	// CreateInOrder is an atomic operation that can be used to enqueue a
+	// request onto a semaphore. In the rest of the comments, we refer to the
+	// resulting key as a "semaphore key".
+	// https://coreos.com/etcd/docs/2.0.8/api.html#atomically-creating-in-order-keys
+	response, err := c.backend.client.CreateInOrder(c.backend.nodePathLock(c.key), "Vault Lock", EtcdLockTTL)
```

Comment:
It could, but I'm not sure what the purpose would be outside of knowing what value queued contenders are trying to set. There isn't an interface method for that.